sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala

```diff
@@ -40,14 +40,43 @@ class ContinuousSuiteBase extends StreamTest {
   protected def waitForRateSourceTriggers(query: StreamExecution, numTriggers: Int): Unit = {
     query match {
       case s: ContinuousExecution =>
-        assert(numTriggers >= 2, "must wait for at least 2 triggers to ensure query is initialized")
-        val reader = s.lastExecution.executedPlan.collectFirst {
-          case ContinuousScanExec(_, _, r: RateStreamContinuousStream, _) => r
-        }.get
+        s.awaitEpoch(0)
 
+        // This is called after waiting first epoch to be committed, so we can just treat
+        // it as partition readers for rate source are already initialized.
+        val firstCommittedTime = System.currentTimeMillis()
         val deltaMs = numTriggers * 1000 + 300
-        while (System.currentTimeMillis < reader.creationTime + deltaMs) {
-          Thread.sleep(reader.creationTime + deltaMs - System.currentTimeMillis)
+        while (System.currentTimeMillis < firstCommittedTime + deltaMs) {
+          Thread.sleep(firstCommittedTime + deltaMs - System.currentTimeMillis)
```

Comment:
I think this could cause an exception if the current time advanced between the check and the sleep, as a negative argument will cause an exception. Maybe:
```
var toWaitMS = firstCommittedTime + deltaMs - System.currentTimeMillis
while (toWaitMS > 0) {
  Thread.sleep(toWaitMS)
  toWaitMS = firstCommittedTime + deltaMs - System.currentTimeMillis
}
```

Reply:
(I guess technically you should use System.nanoTime here but it probably won't matter)

Reply:
You're right we're calling `System.currentTimeMillis` twice which might not be the same, so possible to be negative. Nice finding.

Reply:
Addressed.